### PR TITLE
OT-FrontEndNotificatons | Got firebase to send notifcations to the emulator/flutter application

### LINKED ISSUE #34: OT-FrontEndNotificatons
WHAT TO DO: 
1. Implement a notifications feature in our front-end application.

RESOURCES:
1. https://medium.com/@SebastianEngel/easy-push-notifications-with-flutter-and-firebase-cloud-messaging-d96084f5954f

DONE:
1. You are able to trigger the notification from the firebase console

### PR DESCRIPTION
**Pull request by steve545**
**What I did:**

- I am able to trigger the notification from the firebase console to the emulator.

**How to test:**

- Log into the Firebase web console using the credentials found in this [documentation](https://docs.google.com/document/d/1kqS7te_Rzmt-06OILNQuTDpwEqzwJCJ0VYcxnONbMy4/edit)

- Once logged in, click on the `off-top-firebase` box.

- In the navigation tab on the left, navigate to the `Grow` drop down menu.

- In there, click on `cloud messaging`.

- Launch the Off-Top application.

- Log into the app.

- In the flutter terminal there will be a `FirebaseMessaging token` you need to copy.

image

- In your Firebase window click on `new notification`.

- Enter a `Notification tile `and `Notification Text`.

- Click next and choose the targeting app to be `Off Top Android`.

- Click next and chose the scheduling to be `Now`.

- Click next until you get to step 5, enter the following information to match the photo below

image

- Go back to step 1 shown below 

image

- Click on `send test message` and paste the token from the terminal here and add it by pressing the plus icon.

image

- Click `Test` and when the app is either in the background or foreground the notification will be sent but when the app is in the foreground you should also see the message in the terminal with the data shown in the picture.

**Foreground**

image

**Background**

image

- If you want to test another message notification you can create a new one by clicking `new notification` or you can click on the 3 dots of your last notification and click on duplicate to send another test message by following the same steps above.

### CODE REVIEW

File: lib/pushNotifications.dart [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import 'dart:io';
+
+import 'package:firebase_messaging/firebase_messaging.dart';
+
+class PushNotificationsManager {
```
Comment (tonynguyen98) [inline lib/pushNotifications.dart]: Can you move this to lib/components/pushNotifications/pushNotifications.dart

File: lib/pushNotifications.dart [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  static final PushNotificationsManager _instance =
+      PushNotificationsManager._();
+
+  final FirebaseMessaging _firebaseMessaging = FirebaseMessaging();
+  bool _initialized = false;
+
+  Future<void> init() async {
+    if (!_initialized) {
+      // For iOS request permission first.
```
Comment (tonynguyen98) [inline lib/pushNotifications.dart]: Can you partner up with someone who has iOS and Mac to implement notifications for iOS?

**Review by tonynguyen98 — DISMISSED** [review]
Can you get the iOS notifications working as well?

File: lib/recordingSession.dart [hunk 3]
```diff
@@ -16,12 +17,16 @@ class RecordingPage extends StatefulWidget {
 }
 
 class _RecordingPageState extends State<RecordingPage> {
+  PushNotificationsManager pm;
   final GlobalKey<MeterState> meterState = GlobalKey<MeterState>();
   int yes = 0;
   MyWebSocket ws;
```
Comment (txt020) [inline lib/recordingSession.dart]: Where is int yes being used? Can you remove?

**Review by txt020 — DISMISSED** [review]

File: lib/pushNotifications.dart (hunk 1 above)
Comment (frlzjosh) [reply lib/pushNotifications.dart]: pushNotificationsManager.dart sounds a litttlllle better

File: lib/pushNotifications.dart (hunk 2 above)
Comment (frlzjosh) [reply lib/pushNotifications.dart]: You have me, @upDJ, & @davidsquines to work with

Comment (frlzjosh) [reply lib/pushNotifications.dart]: rename method to something more readible like `initializePushNotificationsManager()`

File: lib/recordingSession.dart [hunk 4]
```diff
@@ -16,12 +17,16 @@ class RecordingPage extends StatefulWidget {
 }
 
 class _RecordingPageState extends State<RecordingPage> {
+  PushNotificationsManager pm;
```
Comment (frlzjosh) [inline lib/recordingSession.dart]: rename variable to something more readible like `pushNotificationsManager` or `notificationsManager`

**Review by frlzjosh — CHANGES_REQUESTED** [review]
It's looking good and it seems like you had to do a few native implementations, which is always a pain. On my end, you have small changes, but it seems like we need to talk about how you can proceed with your task with having the notifications run on ios.

**Review by frlzjosh — APPROVED** [review]
nice man!

File: lib/pushNotifications.dart (hunk 2 above)
Comment (frlzjosh) [reply lib/pushNotifications.dart]: bad news yall, in order for us to send notifications to ios devices we need an Apple Developer account which costs $99 dolllazzz (ayyyyy :zany_face: ) and I don't think it's worth it to purchase now as we're about 68% done. 

So until we have our application working on android that's when we'll focus on deploying onto the apple store and integration the notifications feature for ios devices.

**Pull request merged**